pkg/api/cached.go

```diff
@@ -0,0 +1,99 @@
+package api
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/gadget-inc/dateilager/internal/files"
+	"github.com/gadget-inc/dateilager/internal/key"
+	"github.com/gadget-inc/dateilager/internal/logger"
+	"github.com/gadget-inc/dateilager/internal/pb"
+	"github.com/gadget-inc/dateilager/pkg/client"
+	"golang.org/x/sys/unix"
+)
+
+type Cached struct {
+	pb.UnimplementedCacheServer
+	Client      *client.Client
+	StagingPath string
+	// the current directory holding a fully formed downloaded cache
+	currentDir string
+	// the current version of the cache on disk at currentDir
+	currentVersion int64
+}
+
+func (c *Cached) PopulateDiskCache(ctx context.Context, req *pb.PopulateDiskCacheRequest) (*pb.PopulateDiskCacheResponse, error) {
+	err := requireAdminAuth(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	destination := req.Path
+
+	version, err := c.WriteCache(destination)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.PopulateDiskCacheResponse{Version: version}, nil
+}
+
+// check if the destination exists, and if so, if its writable
+// hardlink the golden copy into this downstream's destination, creating it if need be
+func (c *Cached) WriteCache(destination string) (int64, error) {
+	if c.currentDir == "" {
+		return -1, errors.New("no cache prepared, currentDir is nil")
+	}
+
+	stat, err := os.Stat(destination)
+	if !os.IsNotExist(err) {
+		if err != nil {
+			return -1, fmt.Errorf("failed to stat cache destination %s: %v", destination, err)
+		}
+
+		if !stat.IsDir() {
+			return -1, fmt.Errorf("failed to open cache destination %s for writing -- it is already a file", destination)
+		}
+
+		if unix.Access(destination, unix.W_OK) != nil {
+			return -1, fmt.Errorf("failed to open cache destination %s for writing -- write permission denied", destination)
+		}
+	}
+
+	err = files.HardlinkDir(c.currentDir, destination)
+	if err != nil {
+		return -1, fmt.Errorf("failed to hardlink cache to destination %s: %v", destination, err)
+	}
+	return c.currentVersion, nil
+}
+
+/** Fetch the cache into a spot in the staging dir */
+func (c *Cached) Prepare(ctx context.Context) error {
+	start := time.Now()
+	newDir := path.Join(c.StagingPath, randomString())
+	version, count, err := c.Client.GetCache(ctx, newDir)
```

Comment:
In theory the caching system can support being updated (pulling in a new cache version alongside an existing one) and we could do that to a node after it's been booted up.

But considering the speed of change of caches so far, I think a one shot is fine and new cache versions will appear as nodes cycle in and out.

In the future we likely want to expose some kind of GRPC endpoint to force every node to update it's cache to the latest version.

Reply:
Yep definitely -- the `Prepare` function is built to do that, where it uses a new subdir to stage the cache in over time so it will support the nice pattern of preparing a new one while still serving the old one I think that's an easy improvement but I tried to be disciplined and just get something working for now.

Reply:
So they are actually safe to build within the same directory.

The objects don't overlap (and if they do, DL should be smart enough to not write a hash that already exists) and then once all of the objects are written the version file is updated to contain a new row with the new version.

So you could fold multiple caches into one directory and it's race safe. That's why `availableCacheVersions` is a list.

Reply:
I think it might make sense to still keep the caches in different directories so that we don't have an ever-growing list of files to hardlink into place for each specialization -- cool to leave as is?

Reply:
Yeah I think we'll come back to the updating cache later, we've got a few bridges to cross before nailing that down.